lisp/ess-r-flymake.el

```diff
@@ -0,0 +1,149 @@
+;;
+;; This is distributed in the hope that it will be useful, but WITHOUT
+;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+;; FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+;; for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to the
+;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+;; MA 02110-1301 USA.
+
+
+;;; Commentary:
+;; Flymake is the built-in Emacs package that supports on-the-fly
+;; syntax checking.  This file adds support for this in R-mode by
+;; relying on the lintr package, available on CRAN and currently
+;; hosted at https://github.com/jimhester/lintr.
+
+;; It is enabled by default.
+
+;;; Code:
+
+(eval-when-compile
+  (require 'cl-lib))
+
+(defcustom ess-r-flymake-linters "default_linters"
+  "Default linters to use.
+
+See \"lintr::with_defaults\" for how to customize this."
+  :group 'ess-R
+  :type 'string)
+
+(defcustom ess-r-flymake-lintr-cache t
+  "If non-nil, cache lintr results."
+  :group 'ess-R
+  :type 'string)
+
+(defvar-local ess-r--flymake-proc nil)
+
+(defun ess-r-flymake (report-fn &rest _args)
+  "A Flymake backend for ESS-R modes. Relies on the lintr package.
+
+REPORT-FN is flymake's callback function."
+  (unless (executable-find
+           ;; TODO - check whether lintr package is available
+           "R") (error "Cannot find a suitable R"))
+  ;; If a live process launched in an earlier check was found, that
+  ;; process is killed.  When that process's sentinel eventually runs,
+  ;; it will notice its obsoletion, since it have since reset
+  ;; `ess-r-flymake-proc' to a different value
+  (when (process-live-p ess-r--flymake-proc)
+    (kill-process ess-r--flymake-proc))
+  ;; Save the current buffer, the narrowing restriction, remove any
+  ;; narrowing restriction.
+  (let ((source (current-buffer)))
+    (save-restriction
+      (widen)
+      (setq
+       ess-r--flymake-proc
+       (make-process
+        :name "ess-r-flymake" :noquery t :connection-type 'pipe
+        :buffer (generate-new-buffer " *ess-r-flymake*")
+        :command `("R" "--slave" "--restore" "--no-save" "-e"
+                   ,(eval (concat
+                           "library(lintr);"
+                           ;; commandArgs(TRUE) lets us access
+                           ;; everything after --args as a string:
+                           "try(lint(commandArgs(TRUE)"
+                           ", " "linters = " ess-r-flymake-linters
+                           (when ess-r-flymake-lintr-cache ", cache = TRUE")
+                           "))"))
+                   "--args" ,(eval
+                              ;; text string of the current buffer:
+                              (concat (buffer-substring-no-properties
+                                       (point-min) (point-max)))))
+        :sentinel
+        (lambda (proc _event)
```

Comment:
Would you mind getting this sentinel into a separate function?

Reply:
I think we should leave it as-is, since that's how e.g. python.el and elisp.el do it.